File: hfa-control/src/hfa_control/submission_idempotency.py
"""Tenant-scoped idempotent RUN submission gate."""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
import hashlib
import json
from pathlib import Path
from typing import Any, Mapping

from hfa.config.keys import RedisKey
from hfa.lua.loader import LuaScriptLoader


class SubmissionIdempotencyError(RuntimeError):
    pass


class IdempotencyReservationStatus(str, Enum):
    RESERVED = "RESERVED"
    FINAL_SAME_REQUEST = "FINAL_SAME_REQUEST"
    IN_PROGRESS_SAME_REQUEST = "IN_PROGRESS_SAME_REQUEST"
    DIFFERENT_REQUEST = "DIFFERENT_REQUEST"
    INVALID_EVIDENCE = "INVALID_EVIDENCE"


@dataclass(frozen=True)
class IdempotencyReservation:
    status: IdempotencyReservationStatus
    run_id: str
    task_id: str
    result_json: str = ""


def validate_idempotency_key(value: Any) -> str:
    if not isinstance(value, str):
        raise TypeError(
            "Idempotency-Key must be a string"
        )
    if not value:
        raise ValueError(
            "Idempotency-Key must not be empty"
        )
    if value != value.strip():
        raise ValueError(
            "Idempotency-Key must not contain "
            "surrounding whitespace"
        )
    if len(value) > 128:
        raise ValueError(
            "Idempotency-Key must not exceed 128 characters"
        )
    if any(
        ord(character) < 32
        or ord(character) == 127
        for character in value
    ):
        raise ValueError(
            "Idempotency-Key must not contain control characters"
        )
    return value


def canonical_submission_fingerprint(
    request: Any,
) -> str:
    material = {
        "run_shape": str(
            getattr(request, "run_shape", "") or ""
        ),
        "payload": dict(
            getattr(request, "payload", {}) or {}
        ),
        "agent_type": str(
            getattr(request, "agent_type", "") or ""
        ),
        "priority": getattr(request, "priority", None),
        "estimated_cost_cents": getattr(
            request,
            "estimated_cost_cents",
            None,
        ),
        "preferred_region": str(
            getattr(
                request,
                "preferred_region",
                "",
            )
            or ""
        ),
        "preferred_placement": str(
            getattr(
                request,
                "preferred_placement",
                "",
            )
            or ""
        ),
        "required_capabilities": list(
            getattr(
                request,
                "required_capabilities",
                (),
            )
            or ()
        ),
    }
    encoded = json.dumps(
        material,
        ensure_ascii=False,
        separators=(",", ":"),
        sort_keys=True,
    ).encode("utf-8")
    return hashlib.sha256(encoded).hexdigest()


def idempotency_key_digest(
    idempotency_key: str,
) -> str:
    normalized = validate_idempotency_key(
        idempotency_key
    )
    return hashlib.sha256(
        normalized.encode("utf-8")
    ).hexdigest()


def submission_idempotency_redis_key(
    tenant_id: str,
    idempotency_key: str,
) -> str:
    return RedisKey.submission_idempotency(
        tenant_id,
        idempotency_key_digest(idempotency_key),
    )


def _script_path() -> Path:
    here = Path(__file__).resolve()
    for parent in here.parents:
        for relative in (
            "hfa-core/src/hfa/lua/"
            "submission_idempotency.lua",
            "hfa/lua/submission_idempotency.lua",
            "lua/submission_idempotency.lua",
        ):
            candidate = parent / relative
            if candidate.exists():
                return candidate
    raise FileNotFoundError(
        "submission_idempotency.lua not found"
    )


def _decode(value: Any) -> str:
    if isinstance(value, bytes):
        return value.decode(
            "utf-8",
            errors="strict",
        )
    return str(value) if value is not None else ""


class SubmissionIdempotencyStore:
    """Atomic Redis-backed submission reservation/finalization."""

    def __init__(
        self,
        redis: Any,
        *,
        retention_seconds: int,
        loader: Any | None = None,
    ) -> None:
        if (
            type(retention_seconds) is not int
            or retention_seconds <= 0
        ):
            raise ValueError(
                "retention_seconds must be a positive int"
            )
        self._redis = redis
        self._retention_seconds = retention_seconds
        self._loader = (
            loader
            if loader is not None
            else LuaScriptLoader(
                redis,
                _script_path(),
            )
        )
        self._initialised = False

    @property
    def retention_seconds(self) -> int:
        return self._retention_seconds

    async def initialise(self) -> None:
        if self._initialised:
            return
        await self._loader.load()
        self._initialised = True

    async def reserve(
        self,
        *,
        tenant_id: str,
        idempotency_key: str,
        request_fingerprint: str,
        run_id: str,
        task_id: str,
        owner_token: str,
        created_at_ms: int,
    ) -> IdempotencyReservation:
        await self.initialise()
        result = await self._loader.run(
            num_keys=1,
            keys=[
                submission_idempotency_redis_key(
                    tenant_id,
                    idempotency_key,
                )
            ],
            args=[
                "reserve",
                tenant_id,
                request_fingerprint,
                run_id,
                task_id,
                owner_token,
                str(created_at_ms),
                str(self._retention_seconds),
                "",
            ],
        )
        return self._reservation(result)

    async def finalize(
        self,
        *,
        tenant_id: str,
        idempotency_key: str,
        request_fingerprint: str,
        run_id: str,
        task_id: str,
        owner_token: str,
        result_json: str,
        updated_at_ms: int,
    ) -> None:
        if not result_json:
            raise ValueError(
                "result_json must not be empty"
            )
        await self.initialise()
        result = await self._loader.run(
            num_keys=1,
            keys=[
                submission_idempotency_redis_key(
                    tenant_id,
                    idempotency_key,
                )
            ],
            args=[
                "finalize",
                tenant_id,
                request_fingerprint,
                run_id,
                task_id,
                owner_token,
                str(updated_at_ms),
                str(self._retention_seconds),
                result_json,
            ],
        )
        values = self._values(result)
        if not values or values[0] != "FINALIZED":
            raise SubmissionIdempotencyError(
                "idempotency finalization failed: "
                f"{values!r}"
            )

    async def release(
        self,
        *,
        tenant_id: str,
        idempotency_key: str,
        request_fingerprint: str,
        run_id: str,
        task_id: str,
        owner_token: str,
        updated_at_ms: int,
    ) -> None:
        await self.initialise()
        result = await self._loader.run(
            num_keys=1,
            keys=[
                submission_idempotency_redis_key(
                    tenant_id,
                    idempotency_key,
                )
            ],
            args=[
                "release",
                tenant_id,
                request_fingerprint,
                run_id,
                task_id,
                owner_token,
                str(updated_at_ms),
                str(self._retention_seconds),
                "",
            ],
        )
        values = self._values(result)
        if not values or values[0] != "RELEASED":
            raise SubmissionIdempotencyError(
                "idempotency release failed: "
                f"{values!r}"
            )

    @staticmethod
    def _values(result: Any) -> list[str]:
        if not isinstance(result, (list, tuple)):
            raise SubmissionIdempotencyError(
                "idempotency script returned "
                f"non-sequence result: {result!r}"
            )
        return [_decode(value) for value in result]

    @classmethod
    def _reservation(
        cls,
        result: Any,
    ) -> IdempotencyReservation:
        values = cls._values(result)
        if len(values) < 4:
            raise SubmissionIdempotencyError(
                "idempotency reservation returned "
                f"an incomplete result: {values!r}"
            )
        try:
            status = IdempotencyReservationStatus(
                values[0]
            )
        except ValueError as exc:
            raise SubmissionIdempotencyError(
                "idempotency reservation returned "
                f"unknown status: {values[0]!r}"
            ) from exc
        return IdempotencyReservation(
            status=status,
            run_id=values[1],
            task_id=values[2],
            result_json=values[3],
        )

File: hfa-core/src/hfa/lua/submission_idempotency.lua
-- Tenant-scoped single-task submission idempotency gate.
--
-- KEYS[1] idempotency HASH
-- ARGV:
--   1 action: reserve | finalize | release
--   2 tenant_id
--   3 request_fingerprint
--   4 run_id
--   5 task_id
--   6 owner_token
--   7 timestamp_ms
--   8 ttl_seconds
--   9 result_json

local key = KEYS[1]
local action = ARGV[1]
local tenant_id = ARGV[2]
local fingerprint = ARGV[3]
local run_id = ARGV[4]
local task_id = ARGV[5]
local owner_token = ARGV[6]
local timestamp_ms = ARGV[7]
local ttl_seconds = tonumber(ARGV[8])
local result_json = ARGV[9]

if not ttl_seconds or ttl_seconds <= 0 then
  return {"INVALID_EVIDENCE", "", "", ""}
end

local function key_type()
  local result = redis.call("TYPE", key)
  if type(result) == "table" then
    return result["ok"]
  end
  return result
end

local function current()
  return redis.call(
    "HMGET",
    key,
    "schema_version",
    "tenant_id",
    "request_fingerprint",
    "run_id",
    "task_id",
    "state",
    "owner_token",
    "result_json",
    "created_at_ms",
    "updated_at_ms"
  )
end

local function valid_common(values)
  return (
    values[1] == "1"
    and values[2] == tenant_id
    and values[3] ~= false
    and values[3] ~= ""
    and values[4] ~= false
    and values[4] ~= ""
    and values[5] ~= false
    and values[5] ~= ""
    and values[6] ~= false
    and values[6] ~= ""
  )
end

if action == "reserve" then
  local kind = key_type()
  if kind == "none" then
    redis.call(
      "HSET",
      key,
      "schema_version", "1",
      "tenant_id", tenant_id,
      "request_fingerprint", fingerprint,
      "run_id", run_id,
      "task_id", task_id,
      "state", "IN_PROGRESS",
      "owner_token", owner_token,
      "result_json", "",
      "created_at_ms", timestamp_ms,
      "updated_at_ms", timestamp_ms
    )
    redis.call("EXPIRE", key, ttl_seconds)
    return {"RESERVED", run_id, task_id, ""}
  end

  if kind ~= "hash" then
    return {"INVALID_EVIDENCE", "", "", ""}
  end

  local values = current()
  if not valid_common(values) then
    return {"INVALID_EVIDENCE", "", "", ""}
  end

  if values[3] ~= fingerprint then
    return {
      "DIFFERENT_REQUEST",
      values[4],
      values[5],
      ""
    }
  end

  if values[6] == "FINAL" then
    if values[8] == false or values[8] == "" then
      return {"INVALID_EVIDENCE", values[4], values[5], ""}
    end
    return {
      "FINAL_SAME_REQUEST",
      values[4],
      values[5],
      values[8]
    }
  end

  if values[6] == "IN_PROGRESS" then
    local created_at_ms = tonumber(values[9])
    local updated_at_ms = tonumber(values[10])
    local ttl_remaining = redis.call("TTL", key)
    if (
      values[7] == false
      or values[7] == ""
      or not created_at_ms
      or created_at_ms < 0
      or not updated_at_ms
      or updated_at_ms < created_at_ms
      or ttl_remaining <= 0
    ) then
      return {"INVALID_EVIDENCE", values[4], values[5], ""}
    end
    return {
      "IN_PROGRESS_SAME_REQUEST",
      values[4],
      values[5],
      "",
      tostring(created_at_ms),
      tostring(updated_at_ms),
      tostring(ttl_remaining)
    }
  end

  return {"INVALID_EVIDENCE", values[4], values[5], ""}
end

if action == "finalize" then
  if key_type() ~= "hash" then
    return {"INVALID_EVIDENCE", "", "", ""}
  end
  local values = current()
  if not valid_common(values) then
    return {"INVALID_EVIDENCE", "", "", ""}
  end
  if (
    values[3] ~= fingerprint
    or values[4] ~= run_id
    or values[5] ~= task_id
    or values[6] ~= "IN_PROGRESS"
    or values[7] ~= owner_token
    or result_json == ""
  ) then
    return {"INVALID_EVIDENCE", values[4], values[5], ""}
  end
  redis.call(
    "HSET",
    key,
    "state", "FINAL",
    "result_json", result_json,
    "updated_at_ms", timestamp_ms
  )
  redis.call("EXPIRE", key, ttl_seconds)
  return {"FINALIZED", run_id, task_id, result_json}
end

if action == "release" then
  if key_type() ~= "hash" then
    return {"INVALID_EVIDENCE", "", "", ""}
  end
  local values = current()
  if not valid_common(values) then
    return {"INVALID_EVIDENCE", "", "", ""}
  end
  if (
    values[3] ~= fingerprint
    or values[4] ~= run_id
    or values[5] ~= task_id
    or values[6] ~= "IN_PROGRESS"
    or values[7] ~= owner_token
  ) then
    return {"INVALID_EVIDENCE", values[4], values[5], ""}
  end
  redis.call("DEL", key)
  return {"RELEASED", run_id, task_id, ""}
end

return {"INVALID_EVIDENCE", "", "", ""}
